core/src/main/java/org/elasticsearch/index/query/IndexQueryParserService.java

```diff
@@ -177,15 +181,17 @@ public ParsedQuery parse(QueryParseContext context, BytesReference source) {
         }
     }
 
-    public ParsedQuery parse(String source) throws QueryParsingException {
+    public ParsedQuery parse(String source) throws QueryShardException, QueryShardException {
         XContentParser parser = null;
         try {
             parser = XContentFactory.xContent(source).createParser(source);
             return innerParse(cache.get(), parser);
+        } catch (QueryShardException e) {
+            throw e;
         } catch (QueryParsingException e) {
```

Comment:
After I removed this, some tests in SimpleIndexQueryParsersTest failed, they seem to make assertions in the exception message, so I leave it for now.